File: frontend/src/types/project.ts
export interface Project {
  id: string;
  name: string;
  description: string | null;
  address: string | null;
  client_name: string | null;
  project_number: string | null;
  grundstuecksnr: string | null;
  planverfasser: string | null;
  status: string;
  created_at: string;
  updated_at: string;
}

export interface ProjectCreate {
  name: string;
  description?: string;
  address?: string;
  client_name?: string;
  project_number?: string;
  grundstuecksnr?: string;
  planverfasser?: string;
}

import type { Room } from "./room";

export interface Building {
  id: string;
  project_id: string;
  name: string;
  sort_order: number;
}

export interface BuildingCreate {
  name: string;
  sort_order?: number;
}

export interface BuildingUpdate {
  name?: string;
  sort_order?: number;
}

export interface Floor {
  id: string;
  building_id: string;
  name: string;
  level_number: number | null;
  floor_height_m: number | null;
  sort_order: number;
}

export interface FloorCreate {
  name: string;
  level_number?: number | null;
  floor_height_m?: number | null;
  sort_order?: number;
}

export interface FloorUpdate {
  name?: string;
  level_number?: number | null;
  floor_height_m?: number | null;
  sort_order?: number;
}

export interface Unit {
  id: string;
  floor_id: string;
  name: string;
  unit_type: string | null;
  sort_order: number;
}

export interface UnitCreate {
  name: string;
  unit_type?: string | null;
  sort_order?: number;
}

export interface UnitUpdate {
  name?: string;
  unit_type?: string | null;
  sort_order?: number;
}

// --- Aggregated structure tree ----------------------------------------------
// Mirrors the server-side ``ProjectStructureResponse`` payload. The tree is
// rendered by ``StructurePage`` as collapsible nested cards; keeping the
// nesting on the types (rather than normalizing into a flat list on the
// client) means the UI never has to stitch four separate queries back
// together to redraw after a mutation.

export interface UnitWithRooms extends Unit {
  rooms: Room[];
}

export interface FloorWithUnits extends Floor {
  units: UnitWithRooms[];
}

export interface BuildingWithChildren extends Building {
  floors: FloorWithUnits[];
}

export interface ProjectStructure {
  project_id: string;
  buildings: BuildingWithChildren[];
}

export interface QuickAddResponse {
  project_id: string;
  building_id: string;
  floor_ids: string[];
  unit_ids: string[];
}

File: backend/app/schemas/project.py
"""Pydantic schemas for Project / Building / Floor / Unit and the
aggregate "project structure" tree.

``ProjectStructureResponse`` is what the hierarchical GET endpoint
(``GET /projects/{id}/structure``) returns: a fully-populated
Gebäude → Stockwerk → Einheit → Raum → Öffnung tree so the frontend
can render the collapsible structure page in a single request. We
re-use ``RoomResponse`` / ``OpeningResponse`` from ``schemas.room`` so
the field set stays in lock-step with the flat ``/rooms`` endpoints —
that way the UI never has to branch on "was this fetched via the tree
or via the flat list".
"""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel

from app.schemas.room import RoomResponse


class ProjectCreate(BaseModel):
    name: str
    description: str | None = None
    address: str | None = None
    client_name: str | None = None
    project_number: str | None = None
    grundstuecksnr: str | None = None
    planverfasser: str | None = None


class ProjectUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    address: str | None = None
    client_name: str | None = None
    project_number: str | None = None
    grundstuecksnr: str | None = None
    planverfasser: str | None = None
    status: str | None = None


class ProjectResponse(BaseModel):
    id: UUID
    name: str
    description: str | None
    address: str | None
    client_name: str | None
    project_number: str | None
    grundstuecksnr: str | None
    planverfasser: str | None
    status: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# --- Buildings ---------------------------------------------------------------


class BuildingCreate(BaseModel):
    name: str
    sort_order: int = 0


class BuildingUpdate(BaseModel):
    name: str | None = None
    sort_order: int | None = None


class BuildingResponse(BaseModel):
    id: UUID
    project_id: UUID
    name: str
    sort_order: int

    model_config = {"from_attributes": True}


# --- Floors ------------------------------------------------------------------


class FloorCreate(BaseModel):
    name: str
    level_number: int | None = None
    floor_height_m: float | None = None
    sort_order: int = 0


class FloorUpdate(BaseModel):
    name: str | None = None
    level_number: int | None = None
    floor_height_m: float | None = None
    sort_order: int | None = None


class FloorResponse(BaseModel):
    id: UUID
    building_id: UUID
    name: str
    level_number: int | None
    floor_height_m: float | None
    sort_order: int

    model_config = {"from_attributes": True}


# --- Units -------------------------------------------------------------------


class UnitCreate(BaseModel):
    name: str
    unit_type: str | None = None
    sort_order: int = 0


class UnitUpdate(BaseModel):
    name: str | None = None
    unit_type: str | None = None
    sort_order: int | None = None


class UnitResponse(BaseModel):
    id: UUID
    floor_id: UUID
    name: str
    unit_type: str | None
    sort_order: int

    model_config = {"from_attributes": True}


# --- Aggregated structure tree -----------------------------------------------
#
# Pydantic evaluates the nested models bottom-up: Unit holds Rooms,
# Floor holds Units, Building holds Floors. Each tier sets
# ``from_attributes=True`` so we can return SQLAlchemy instances
# directly from the endpoint (``selectinload`` in the query eagerly
# loads the whole tree, so no lazy-load calls fire at serialization
# time).


class UnitWithRooms(BaseModel):
    id: UUID
    floor_id: UUID
    name: str
    unit_type: str | None
    sort_order: int
    rooms: list[RoomResponse] = []

    model_config = {"from_attributes": True}


class FloorWithUnits(BaseModel):
    id: UUID
    building_id: UUID
    name: str
    level_number: int | None
    floor_height_m: float | None
    sort_order: int
    units: list[UnitWithRooms] = []

    model_config = {"from_attributes": True}


class BuildingWithChildren(BaseModel):
    id: UUID
    project_id: UUID
    name: str
    sort_order: int
    floors: list[FloorWithUnits] = []

    model_config = {"from_attributes": True}


class ProjectStructureResponse(BaseModel):
    """Single-shot payload for the collapsible project-tree UI."""

    project_id: UUID
    buildings: list[BuildingWithChildren] = []


# --- Quick-add ---------------------------------------------------------------


class QuickAddResponse(BaseModel):
    """Returned by the "Schnell-Anlage: Einfamilienhaus" endpoint.

    Echoes back the IDs of everything that was just created so the
    frontend can scroll to / expand the new nodes without a re-fetch
    round-trip.
    """

    project_id: UUID
    building_id: UUID
    floor_ids: list[UUID] = []
    unit_ids: list[UUID] = []
